apm-agent-plugins/apm-micrometer-plugin/src/main/java/co/elastic/apm/agent/micrometer/MicrometerMetricsReporter.java

```diff
@@ -24,81 +24,178 @@
-        for (MeterRegistry registry : meterRegistries) {
+        long metricsIntervalMs = tracer.getConfig(ReporterConfiguration.class).getMetricsIntervalMs();
+        if (metricsIntervalMs == 0) {
+            //by checking here rather than at thread creation/non-creation,
+            //the interval can be dynamically updated and any new value will apply
+            return;
+        }
+
+        //non-step metrics are reported every metricsInterval,
+        //step metrics are reported every step interval (step defined by the registry)
+        boolean reportNonStepMetrics = false;
+        if((now - lastMetricIntervalTimestamp) >= metricsIntervalMs) {
+            reportNonStepMetrics = true;
+            //non-atomic, but this is used single-threaded, it's only
+            //volatile for tests to be able to reset the value
+            lastMetricIntervalTimestamp += metricsIntervalMs;
+        }
+
+        Iterator<Map.Entry<MeterRegistry, Step>> registriesIterator = meterRegistries.iterator();
+        final Set<MeterRegistry> currentlyReportableRegistries = new HashSet<>();
+        while(registriesIterator.hasNext()) {
+            Map.Entry<MeterRegistry, Step> meterRegistryStepEntry = registriesIterator.next();
+            MeterRegistry meterRegistry = meterRegistryStepEntry.getKey();
+            Step meterRegistryStep = meterRegistryStepEntry.getValue();
+            if (meterRegistryStep.isStep()) {
+                logger.debug("Evaluating whether to report: step {}, interval(ms) {}, meterRegistry {}", meterRegistryStep.lastStepCount ,meterRegistryStep.stepInMs, meterRegistry);
+                if (meterRegistryStep.shouldReportNow(now)) {
+                    currentlyReportableRegistries.add(meterRegistry);
+                    meterRegistryStep.incrementToNextStep(now);
+                }
+            } else if (reportNonStepMetrics) {
+                currentlyReportableRegistries.add(meterRegistry);
+            }
+        }
+
+        MeterMapConsumer meterConsumer = MeterMapConsumer.INSTANCE.reset(tracer.getConfig(ReporterConfiguration.class).getDisableMetrics());
+        for (MeterRegistry registry : currentlyReportableRegistries) {
             registry.forEachMeter(meterConsumer);
         }
         logger.debug("Reporting {} meters", meterConsumer.meters.size());
-        for (JsonWriter serializedMetricSet : serializer.serialize(meterConsumer.meters, timestamp)) {
+        for (JsonWriter serializedMetricSet : serializer.serialize(meterConsumer.meters, now * 1000)) {
             reporter.report(serializedMetricSet);
         }
     }
 
+    private long getStep(MeterRegistry meterRegistry) {
+        if (meterRegistry.config() instanceof StepRegistryConfig) {
+            return ((StepRegistryConfig) (meterRegistry.config())).step().toMillis();
+        }
+        if (meterRegistry instanceof SimpleMeterRegistry) {
+            SimpleConfig config = configMap.get(meterRegistry);
```

Comment:
Our current documentation states the following:
```
Attach the agent, and you’re done! The agent automatically detects all MeterRegistry instances and reports all metrics to APM Server (in addition to where they originally report). When attaching the agent after the application has already started, the agent detects a MeterRegistry when calling any public method on it.
```

This is not 100% true after this change, as the behaviour can differ between attachment on startup vs later.

If attaching after startup, we might miss the construction of the `SimpleMeterRegistry` here.

Is this known already?
An alternative would be to instrument some other side-effect free method on `SimpleMeterRegistry` and to call it ourselves, but this also isn't guaranteed to work for subclasses of `SimpleMeterRegistry` unless we directly call the method via MethodHandles.

This might be too much effort and we might as well just add this limitation to the docs instead.

Reply:
excellent point. I'll revisit how we get the config

Reply:
I've added support for attaching after startup. Unfortunately you can't redefine methods after they've been loaded, and there is only one public method in SimpleMeterRegistry to use to trigger getting the private config, and that is only available from 1.9.0 of micrometer, so the post-attach mechanism only works in that scenario. Which I think is fine